### ISSUE TITLE
Make accessible to new C++ programmers

### ISSUE THREAD

**Issue by danny305** [rbock/sqlpp11#283]
I am new to C++ (<1 year) and I am learning how to use external libraries. I have not been able to find online or in this repo how to even begin using this library. I had a friend mention that this library is a header only library but everything I see talks about building it. So once i am done building it how do I begin using it? I can't find a resource that shows me a quick example on what header files need to be imported to do certain things. How do I even compile a cpp script that uses this library? do I use the -I flag or the -L and -l flag.  I am having a hard time just creating a simple sqlite3 db. In the youtube videos found, all the examples exclude the header files imported nor do they show how to compile the code. 

 I am accustomed to python where I can just pip install sqlalchemy. import the module and just point it to a sqlite3 file in my directory and begin populating it. I can't find any documentation showing me how to do that. 

In summary, this library is inaccessible to programmers who are not seasoned C++ veterans. Also, I do not know if this should be posted here or stack overflow.

 Any help and clarification at all would be much appreciated.

**Comment by rbock** [comment]
Hi Danny,

Writing good documentation is probably the hardest part of software engineering. Let me try to clear a few things up:

sqlpp11 itself is header only. You don't need to compile it to install it. If you want to run any of the test programs, you will need to compile those, of course. And if you want to use it in your software, then the library (or parts of it) will be compiled as part of your code, wherever you include it (depending on where you put it, you will need to indicate that location to the compiler, e.g. with the `-I` flag for gcc)

sqlpp11 itself is agnostic of the actual database you use (e.g. sqlite, mysql, postgresql, oracle, ...). In order to connect it to a concrete database, you need to use a second library, the connector library. These are located in separate repositories, e.g. https://github.com/rbock/sqlpp11-connector-sqlite3 for sqlite.

The connector libraries are not header only. You need to compile them to use them. To use the connector libraries, you need to indicate header locations (`-I`), library locations (`-L`) and the library name (`-l`). The connector libraries (as well as sqlpp11 itself) come with test programs, which insert into, update, or select/delete from database tables. cmake generates the respective compiler and linker commands with the respective flags.

sqlpp11 (and its connector libraries) also know nothing about your table structure. The library needs to know about the table structure in order to tell you at compile time if you are doing something wrong (like selecting a column that does not exist, or trying to interpret a value as text, when it is numeric instead).

The need to know the table structure at compile time might be the greatest mental stumbling block for newcomers to the library. Most other libraries (including alchemy) analyse table structures at runtime. However, if the compiler knows the table structure at compile time, it can prevent whole classes of bugs from appearing in your tests or in production. 

For this purpose you need to generate header files from your SQL table definitions, using scripts/ddl2cpp script (for sqlite, there is an alternative called sqlite2cpp.py).

The `Sample.h` include file that is used in the test programs is generated this way.

In a project, you would typically store your DDL files as source code in your repository and generate the table header files as part of the build process.

That being said, there is an alternative to the header file generation: @niXman created a code generator inside of sqlpp11 that can generate the table types without includes, simply from hand-written code, see `tests/Ppgen.cpp`.

I hope this helps?

As stated above, writing good documentation is hard. Please feel free to ask more questions.

If you succeed (I am sure you will), I would appreciate if you would write a 'How to get started' guide.


Best,

Roland

**Comment by rbock** [comment]
I am compiling a wish list out of long-standing issues. Better documentation is certainly one of the items (and probably the one I will be working on myself the most, when I find the time).

Closing this ticket.

**Comment by danny305** [comment]
Roland,

Sorry for the late response. I would love to help you make some
documentation tailored toward more inexperienced C++ developers, like
myself. It would be a great way for me to become familiar with your library
and grow my skills in C++.

I have blocked out time between today and tomorrow to go through your
suggestions and attempt to implement them. If you wish to work together,
towards the end of this week I can have a list questions for you (or we can
just have on going conversations as issues come up) and I can begin
documenting what I did to overcome some of the hurdles I encountered
(Create a markdown file or a well commented Cpp file? Let me know the
format you would prefer me to document my experience.)

Let me provide you some background about myself and on why I want to use
your library and what my overarching goal is:

I am a graduate student at the University of Texas in the chemistry
department. I am trying to build a data pipeline in order to train neural
networks on protein crystal structure data on the fly (generate the data
from the crystal structure as the net is training). I am building a
database to house all the files that store the positions and properties of
all the atoms in a protein (protein crystal structure data).  The goal is
to pull these files from the database and generate, on the fly, the data
representation that a neural net will need to learn. We do not want to save
these encodings, we just want to generate them as the net needs them and
delete them once it is done training.

Today I will spend some time trying to connect to a sqlite3 db and grab
rows from a table I made on Sunday. I got the posgresql and sqlite3
connectors built. I managed to build the initial table for the DB via
sqlalchemy on Sunday. However, the data pipeline I am writing is all in C++
(the python pipeline takes nearly 1 week to run and requires us to store TB
of data encodings, which is why I am rewriting in C++). So the first thing
I need to do with your library (and connector library ) is connect to the
DB (more specifically the table I made) and begin reading rows to begin the
data encoding generation. (The rows consist of 3 ASCII files that have been
expanded into their own column. VARCHAR 15000000) The goal is to read in a
row, grab the 3 ASCII file and parse them to begin generating encoded
representation for machine learning.


I look forward to working with you. Going forward do you prefer to
communicate via email or should we move this back to Github?

Danny

**Comment by rbock** [comment]
Hi Danny,

That sound like an exciting project!

Regarding the documentation, I would prefer markdown with embedded code examples. Accompanying example "projects" would be awesome.

Looking forward to your first draft. The cooperation has to be asynchronous most of the time as I am living in Munich. 

For specific questions, I would suggest opening individual issues.

Cheers,

Roland

**Comment by danny305** [comment]
For my first batch of questions:

I started the day by going through the wiki and finding an example that
would allow me to connect to the db and select columns from a table.
So I decided to try and replicate the code for SampleTest.cpp in the
sqlite3-connector test directory.

This is what I have:
[image: image.png]
I have spent the last few hours trying to learn how to compile this code. I
can compile if I comment out the db(config) line but as soon as I
incorporate the line I get linking errors. So I went to my /usr/local/lib/
and this is the only sqlite3 sqlpp library I have:
[image: image.png]

So in my cmakelists.txt file this is what I have:
[image: image.png]

However, when I run make:

make

[  6%] Linking CXX executable openDB

Undefined symbols for architecture x86_64:

  "_sqlite3_bind_blob", referenced from:

      sqlpp::sqlite3::prepared_statement_t::_bind_blob_parameter(unsigned
long, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> >
const*, bool) in libsqlpp11-connector-sqlite3.a(prepared_statement.cpp.o)

  "_sqlite3_bind_double", referenced from:


sqlpp::sqlite3::prepared_statement_t::_bind_floating_point_parameter(unsigned
long, double const*, bool) in
libsqlpp11-connector-sqlite3.a(prepared_statement.cpp.o)

  "_sqlite3_bind_int", referenced from:


sqlpp::sqlite3::prepared_statement_t::_bind_boolean_parameter(unsigned
long, signed char const*, bool) in
libsqlpp11-connector-sqlite3.a(prepared_statement.cpp.o)

  "_sqlite3_bind_int64", referenced from:


sqlpp::sqlite3::prepared_statement_t::_bind_integral_parameter(unsigned
long, long long const*, bool) in
libsqlpp11-connector-sqlite3.a(prepared_statement.cpp.o)

  "_sqlite3_bind_null", referenced from:


sqlpp::sqlite3::prepared_statement_t::_bind_boolean_parameter(unsigned
long, signed char const*, bool) in
libsqlpp11-connector-sqlite3.a(prepared_statement.cpp.o)


sqlpp::sqlite3::prepared_statement_t::_bind_floating_point_parameter(unsigned
long, double const*, bool) in
libsqlpp11-connector-sqlite3.a(prepared_statement.cpp.o)


sqlpp::sqlite3::prepared_statement_t::_bind_integral_parameter(unsigned
long, long long const*, bool) in
libsqlpp11-connector-sqlite3.a(prepared_statement.cpp.o)

      sqlpp::sqlite3::prepared_statement_t::_bind_text_parameter(unsigned
long, std::__1::basic_string<char, std::__1::char_traits<char>,
std::__1::allocator<char> > const*, bool) in
libsqlpp11-connector-sqlite3.a(prepared_statement.cpp.o)

      sqlpp::sqlite3::prepared_statement_t::_bind_date_parameter(unsigned
long, std::__1::chrono::time_point<std::__1::chrono::system_clock,
std::__1::chrono::duration<int, std::__1::ratio<86400l, 1l> > > const*,
bool) in libsqlpp11-connector-sqlite3.a(prepared_statement.cpp.o)


sqlpp::sqlite3::prepared_statement_t::_bind_date_time_parameter(unsigned
long, std::__1::chrono::time_point<std::__1::chrono::system_clock,
std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000l> > >
const*, bool) in libsqlpp11-connector-sqlite3.a(prepared_statement.cpp.o)

      sqlpp::sqlite3::prepared_statement_t::_bind_blob_parameter(unsigned
long, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> >
const*, bool) in libsqlpp11-connector-sqlite3.a(prepared_statement.cpp.o)

      ...

  "_sqlite3_bind_text", referenced from:


sqlpp::sqlite3::prepared_statement_t::_bind_floating_point_parameter(unsigned
long, double const*, bool) in
libsqlpp11-connector-sqlite3.a(prepared_statement.cpp.o)

      sqlpp::sqlite3::prepared_statement_t::_bind_text_parameter(unsigned
long, std::__1::basic_string<char, std::__1::char_traits<char>,
std::__1::allocator<char> > const*, bool) in
libsqlpp11-connector-sqlite3.a(prepared_statement.cpp.o)

      sqlpp::sqlite3::prepared_statement_t::_bind_date_parameter(unsigned
long, std::__1::chrono::time_point<std::__1::chrono::system_clock,
std::__1::chrono::duration<int, std::__1::ratio<86400l, 1l> > > const*,
bool) in libsqlpp11-connector-sqlite3.a(prepared_statement.cpp.o)


sqlpp::sqlite3::prepared_statement_t::_bind_date_time_parameter(unsigned
long, std::__1::chrono::time_point<std::__1::chrono::system_clock,
std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000l> > >
const*, bool) in libsqlpp11-connector-sqlite3.a(prepared_statement.cpp.o)

  "_sqlite3_changes", referenced from:


sqlpp::sqlite3::connection::run_prepared_execute_impl(sqlpp::sqlite3::prepared_statement_t&)
in libsqlpp11-connector-sqlite3.a(connection.cpp.o)

      sqlpp::sqlite3::connection::execute(std::__1::basic_string<char,
std::__1::char_traits<char>, std::__1::allocator<char> > const&) in
libsqlpp11-connector-sqlite3.a(connection.cpp.o)

      sqlpp::sqlite3::connection::update_impl(std::__1::basic_string<char,
std::__1::char_traits<char>, std::__1::allocator<char> > const&) in
libsqlpp11-connector-sqlite3.a(connection.cpp.o)


sqlpp::sqlite3::connection::run_prepared_update_impl(sqlpp::sqlite3::prepared_statement_t&)
in libsqlpp11-connector-sqlite3.a(connection.cpp.o)

      sqlpp::sqlite3::connection::remove_impl(std::__1::basic_string<char,
std::__1::char_traits<char>, std::__1::allocator<char> > const&) in
libsqlpp11-connector-sqlite3.a(connection.cpp.o)


sqlpp::sqlite3::connection::run_prepared_remove_impl(sqlpp::sqlite3::prepared_statement_t&)
in libsqlpp11-connector-sqlite3.a(connection.cpp.o)

  "_sqlite3_close", referenced from:


sqlpp::sqlite3::detail::connection_handle::connection_handle(sqlpp::sqlite3::connection_config)
in libsqlpp11-connector-sqlite3.a(connection_handle.cpp.o)

      sqlpp::sqlite3::detail::connection_handle::~connection_handle() in
libsqlpp11-connector-sqlite3.a(connection_handle.cpp.o)

  "_sqlite3_column_blob", referenced from:

      sqlpp::sqlite3::bind_result_t::_bind_blob_result(unsigned long,
unsigned char const**, unsigned long*) in
libsqlpp11-connector-sqlite3.a(bind_result.cpp.o)

  "_sqlite3_column_bytes", referenced from:

      sqlpp::sqlite3::bind_result_t::_bind_text_result(unsigned long, char
const**, unsigned long*) in
libsqlpp11-connector-sqlite3.a(bind_result.cpp.o)

      sqlpp::sqlite3::bind_result_t::_bind_blob_result(unsigned long,
unsigned char const**, unsigned long*) in
libsqlpp11-connector-sqlite3.a(bind_result.cpp.o)

  "_sqlite3_column_double", referenced from:

      sqlpp::sqlite3::bind_result_t::_bind_floating_point_result(unsigned
long, double*, bool*) in libsqlpp11-connector-sqlite3.a(bind_result.cpp.o)

  "_sqlite3_column_int", referenced from:

      sqlpp::sqlite3::connection::get_default_isolation_level() in
libsqlpp11-connector-sqlite3.a(connection.cpp.o)

      sqlpp::sqlite3::bind_result_t::_bind_boolean_result(unsigned long,
signed char*, bool*) in libsqlpp11-connector-sqlite3.a(bind_result.cpp.o)

  "_sqlite3_column_int64", referenced from:

      sqlpp::sqlite3::bind_result_t::_bind_integral_result(unsigned long,
long long*, bool*) in libsqlpp11-connector-sqlite3.a(bind_result.cpp.o)

  "_sqlite3_column_text", referenced from:

      sqlpp::sqlite3::bind_result_t::_bind_floating_point_result(unsigned
long, double*, bool*) in libsqlpp11-connector-sqlite3.a(bind_result.cpp.o)

      sqlpp::sqlite3::bind_result_t::_bind_text_result(unsigned long, char
const**, unsigned long*) in
libsqlpp11-connector-sqlite3.a(bind_result.cpp.o)

      sqlpp::sqlite3::bind_result_t::_bind_date_result(unsigned long,
std::__1::chrono::time_point<std::__1::chrono::system_clock,
std::__1::chrono::duration<int, std::__1::ratio<86400l, 1l> > >*, bool*) in
libsqlpp11-connector-sqlite3.a(bind_result.cpp.o)

      sqlpp::sqlite3::bind_result_t::_bind_date_time_result(unsigned long,
std::__1::chrono::time_point<std::__1::chrono::system_clock,
std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000l> > >*,
bool*) in libsqlpp11-connector-sqlite3.a(bind_result.cpp.o)

  "_sqlite3_column_type", referenced from:

      sqlpp::sqlite3::bind_result_t::_bind_boolean_result(unsigned long,
signed char*, bool*) in libsqlpp11-connector-sqlite3.a(bind_result.cpp.o)

      sqlpp::sqlite3::bind_result_t::_bind_floating_point_result(unsigned
long, double*, bool*) in libsqlpp11-connector-sqlite3.a(bind_result.cpp.o)

      sqlpp::sqlite3::bind_result_t::_bind_integral_result(unsigned long,
long long*, bool*) in libsqlpp11-connector-sqlite3.a(bind_result.cpp.o)

      sqlpp::sqlite3::bind_result_t::_bind_date_result(unsigned long,
std::__1::chrono::time_point<std::__1::chrono::system_clock,
std::__1::chrono::duration<int, std::__1::ratio<86400l, 1l> > >*, bool*) in
libsqlpp11-connector-sqlite3.a(bind_result.cpp.o)

      sqlpp::sqlite3::bind_result_t::_bind_date_time_result(unsigned long,
std::__1::chrono::time_point<std::__1::chrono::system_clock,
std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000l> > >*,
bool*) in libsqlpp11-connector-sqlite3.a(bind_result.cpp.o)

  "_sqlite3_errmsg", referenced from:

      sqlpp::sqlite3::(anonymous
namespace)::prepare_statement(sqlpp::sqlite3::detail::connection_handle&,
std::__1::basic_string<char, std::__1::char_traits<char>,
std::__1::allocator<char> > const&) in
libsqlpp11-connector-sqlite3.a(connection.cpp.o)

      sqlpp::sqlite3::(anonymous
namespace)::execute_statement(sqlpp::sqlite3::detail::connection_handle&,
sqlpp::sqlite3::detail::prepared_statement_handle_t&) in
libsqlpp11-connector-sqlite3.a(connection.cpp.o)


sqlpp::sqlite3::detail::connection_handle::connection_handle(sqlpp::sqlite3::connection_config)
in libsqlpp11-connector-sqlite3.a(connection_handle.cpp.o)

      sqlpp::sqlite3::detail::connection_handle::~connection_handle() in
libsqlpp11-connector-sqlite3.a(connection_handle.cpp.o)

  "_sqlite3_finalize", referenced from:


sqlpp::sqlite3::detail::prepared_statement_handle_t::~prepared_statement_handle_t()
in libsqlpp11-connector-sqlite3.a(connection.cpp.o)

  "_sqlite3_last_insert_rowid", referenced from:

      sqlpp::sqlite3::connection::insert_impl(std::__1::basic_string<char,
std::__1::char_traits<char>, std::__1::allocator<char> > const&) in
libsqlpp11-connector-sqlite3.a(connection.cpp.o)


sqlpp::sqlite3::connection::run_prepared_insert_impl(sqlpp::sqlite3::prepared_statement_t&)
in libsqlpp11-connector-sqlite3.a(connection.cpp.o)

      sqlpp::sqlite3::connection::last_insert_id() in
libsqlpp11-connector-sqlite3.a(connection.cpp.o)

  "_sqlite3_open_v2", referenced from:


sqlpp::sqlite3::detail::connection_handle::connection_handle(sqlpp::sqlite3::connection_config)
in libsqlpp11-connector-sqlite3.a(connection_handle.cpp.o)

  "_sqlite3_prepare_v2", referenced from:

      sqlpp::sqlite3::(anonymous
namespace)::prepare_statement(sqlpp::sqlite3::detail::connection_handle&,
std::__1::basic_string<char, std::__1::char_traits<char>,
std::__1::allocator<char> > const&) in
libsqlpp11-connector-sqlite3.a(connection.cpp.o)

  "_sqlite3_reset", referenced from:

      sqlpp::sqlite3::prepared_statement_t::_reset() in
libsqlpp11-connector-sqlite3.a(prepared_statement.cpp.o)

  "_sqlite3_step", referenced from:

      sqlpp::sqlite3::(anonymous
namespace)::execute_statement(sqlpp::sqlite3::detail::connection_handle&,
sqlpp::sqlite3::detail::prepared_statement_handle_t&) in
libsqlpp11-connector-sqlite3.a(connection.cpp.o)

      sqlpp::sqlite3::bind_result_t::next_impl() in
libsqlpp11-connector-sqlite3.a(bind_result.cpp.o)

ld: symbol(s) not found for architecture x86_64

clang: error: linker command failed with exit code 1 (use -v to see
invocation)

make[2]: *** [openDB] Error 1

make[1]: *** [CMakeFiles/openDB.dir/all] Error 2

make: *** [all] Error 2



I have spent the last few hours learning more about linking and more about
cmake. These things are not intuitive and I think there should be clear
documentation on how to build and link sqlpp11 with different connector
with and without cmake.

 As you can see from a screenshot above, I am beginning with sqlite3 and
plan on redoing everything I learn to do with a production postgresdb. I
will document how to get up and running with both.

My next set of questions are going to be over how to provide sqlpp11 the
schema to my already built table. I am trying to wrap my head around some
of the scripts where they do that and I just do not understand what is
going on there at all. I will spend more time trying to make sense of it.

Look forward to your comments!

Danny

**Comment by rbock** [comment]
Hi Danny,

Thanks for all the details. This is educational for me, too. I am getting a better understanding of your documentation needs.

The images you send (this is still going through github), are being dropped. Code as images is sub-optimal anyway (copy&paste would be difficult for me). Please use github's markdown for code.

Did you solve the linker issues in the end? If you want to use the library, the simplest way is to install it and then tell the compiler to link to it, like 

```
g++ Sample.cpp -I<install-prefix>/include -L<install-prefix>/lib -lsqlpp11-connector-sqlite3 -lsqlite3
```

Cheers,

Roland

**Comment by danny305** [comment]
I just realized that this is still going through GitHub. I will learn how
to do GitHub markdown for code and send you future questions that way. I
had a friend show me some new cmake functionality this morning and I just
got it to compile! (I just started using cmake 2 weeks ago... still
learning).

I will email you later today with some GitHub markdown on what I figured
out and would like your feedback on it so I can understand what I am doing.

Danny

**Comment by Erroneous1** [comment]
Learning (at least basic) CMake is well worth it for a lot of C++ libraries. It makes sqlpp11 and others easier to use. For instance, if you installed sqlpp11, you can use it with:
```cmake
find_package(Sqlpp11 REQUIRED)
#use add_executable or add_library to make a target named MyTarget
target_link_libraries(MyTarget PRIVATE sqlpp11)
```
And if you don't want to install, you can make a libs directory in your project and add sqlpp11 there and use cmake to import it:
```cmake
add_subdirectory(libs/sqlpp11)
target_link_libraries(MyTarget PRIVATE sqlpp11)
```
BTW easy markdown for code is surround small segments of code with `` ` `` (back tick) or for multi line start with 3 `` ``` `` and optionally what language it is (ie `` ```cmake`` or `` ```c++``) on a line by itself and then at the end of the code block do another `` ``` ``.

<pre>```c++
#include <iostream>
int main()
{
  std::cout << "hello world\n";
}
```</pre>
becomes
```c++
#include <iostream>
int main()
{
  std::cout << "hello world\n";
}
```
Don't forget to check the Preview tab.

**Comment by danny305** [comment]
So I am trying to create the table header file using sqlite2cpp.py.
Here is my sql file: (created directly from the sqlite3 command line)
```sql
CREATE TABLE pdb_db
(
id INTEGER NOT NULL,
pdb_id VARCHAR(4) NOT NULL,
file VARCHAR(15000000),
pqr VARCHAR(15000000),
sasa VARCHAR(15000000),
creation_timestamp DATETIME DEFAULT CURRENT_TIMESTAMP,
PRIMARY KEY (id),
CONSTRAINT "3_uniq_filenames" UNIQUE (file, pqr, sasa),
UNIQUE (pdb_id)
);
```

Here is the command I run on the on command line to do it:
```bash

$  ./sqlite2cpp.py ../../../src/pdb_table.sql  ../../../include/ PDB
```

This is the error I get:

```bash

Error: datatype "varchar(4)"" is not supported.

Error: datatype "varchar(15000000)"" is not supported.

Error: datatype "varchar(15000000)"" is not supported.

Error: datatype "varchar(15000000)"" is not supported.

Error: datatype "datetime"" is not supported.

Error: unsupported datatypes.

Possible solutions:

A) Implement this datatype (examples: sqlpp11/data_types)

B) Extend/upgrade ddl2cpp (edit types map)

C) Raise an issue on github
```
I have tried editing my ddl and nothing changes the error message.

I don't understand because in the ddl2cpp_sample_good.sql file the table
uses `varchar(255)`.

Do I need to go into sqlpp11/data_types  and create an entire directory
with all the column_operator.h, concat.h, data_type.h, like.h ....... to
get my date time and varchar to work??

I am very confused on how to proceed. I my real goal is to be able to loop
through my tables rows and grab the strings stored in the file column and
pass it to a pipeline in C++.

Danny

**Comment by rbock** [comment]
While the error message is probably not ideal (@egorpugin might want to chime in here), this is actually an issue with your table definition.

Although sqlite3 "understands" what you mean by varchar(15000000), for instance, it only has very few data types: INTEGER, REAL, and TEXT (see https://www.sqlite.org/datatype3.html).

AFAICT, your table definition should look like this for sqlite3

```sql
CREATE TABLE pdb_db
(
id INTEGER NOT NULL,
pdb_id TEXT NOT NULL,
file TEXT,
pqr TEXT,
sasa TEXT,
creation_timestamp TEXT DEFAULT CURRENT_TIMESTAMP,
PRIMARY KEY (id),
CONSTRAINT "3_uniq_filenames" UNIQUE (file, pqr, sasa),
UNIQUE (pdb_id)
);
```

**Comment by danny305** [comment]
Wow that worked!!! I lost hope today. Not going to lie to you.

I will spend sometime tomorrow reading through that link you sent me and
trying to move forward on selecting rows from the db.

So changing the varchar to text shouldn't make much of a difference in
functionality. Essentially, text is just a string that is limitless (which
is actually the functionality that I want). (correct me if Im wrong)
However, changing my date time column to a text.... Should I expect this to
cause any side effects downstream?

Also, are these changes applicable when creating this header file for this
same table for a postgres backend? (We ultimately want to implement this in
postgres but are developing in sqlite3.) I know I used the sqlite2cpp
script. Should I expect a difference in behavior when using the dll2cpp
script?

Danny

**Comment by rbock** [comment]
ddl2cpp will work with your original ddl.

As for the date time column, try inserting something into your original table definition in sqlite and then look at the timestamp, e.g. 

```sql
insert into pdb_db (id, pdb_id) VALUES (1, "xxx");
select creation_timestamp, typeof(creation_timestamp) from pdb_db;
```

If you want to run stuff with postgresql in production, why don't you use postgresql in development? Databases do have differences (as you just learned). Migrating from one to another is not trivial.

Best,

Roland

**Comment by danny305** [comment]
Roland,

I have gotten the library successfully working with sqlite3. I am able to
select a row, grab the file column, and pass it to my library that build a
protein object. Next step is for me to reproduce this with a postgres db. I
will start from the very beginning and try to get that up. Classes have
just started so my progress will be much slower.

However, I am thinking within the next 2-3 weeks I will have a rough draft
of documentation for you. I will start from git cloning the libraries,
installation, and curating a cmake file to get a simple example compiling
(using sqlite3 and postgres). I think writing this document will really
help me learn why each step is necessary and what exactly each step does. I
look forward on sharing it with you and you explaining in detail a lot of
stuff that I have superficial understanding on.

Danny

**Comment by rbock** [comment]
Hi Danny,

I am happy to read you make progress, and I am looking forward to your first draft :-)

Best,

Roland

**Comment by danny305** [comment]
Quick question:

How can I access the return of this selection (it returns 1 row) without me
having to loop through it and grab it from the row:

```C++
const auto pdb_db =  PDB::PdbDb {};
auto selection = db(select(pdb_db.id,
pdb_db.file).from(pdb_db).where(pdb_db.pdbId == pdb_id));
for (const auto& row : selection)
{
stream << row.file;
}
```

I want to do something like this:
```C++
const auto pdb_db =  PDB::PdbDb {};
auto selection = db(select(pdb_db.id,
pdb_db.file).from(pdb_db).where(pdb_db.pdbId == pdb_id));
stream << selection.file;
```
I have not been able to find an example of this in any of the example
scripts.

Danny

**Comment by rbock** [comment]
Hi,

You can check whether there are no rows or there is at least one row by using `.empty()`.
If you know for sure that there is at least one row you can use `.front()` to get a reference to the first row.

```c++
if (not selection.emty())
{
   stream << selection.front().file;
}
```

Best,

Roland
